Added project structure and basic implementation

enescualexandru opened the pull request.

enescualexandru committed:
Added project structure and basic implementation

TheBestPessimist commented on src/main/kotlin/com/shortener/UrlShortenerApp.kt:
```diff
@@ -0,0 +1,11 @@
+
+fun main(args: Array<String>) {
+	runApplication<ShortenerApplication>(*args)
+}
```
general note for all files: please enable in the formatter "newline at the end of file"

enescualexandru replied:
Enabled. Also corrected the files

TheBestPessimist commented on src/main/kotlin/com/shortener/UrlShortenerApp.kt:
```diff
@@ -0,0 +1,11 @@
+@SpringBootApplication
+class ShortenerApplication
+
+fun main(args: Array<String>) {
```
style: in kotlin, one-liners are written as follows:

```kotlin
fun main(args: Array<String>) = runApplication<ShortenerApplication>(*args)
```

enescualexandru replied:
Tried, looks like maven is not considering this as the "main" class anymore

TheBestPessimist replied:
This makes no sense whatsoever. Why would maven not recognize this? 
Single expression functions (https://kotlinlang.org/docs/functions.html#single-expression-functions) have nothing to do with maven.

enescualexandru replied:
Looks like it is the `spring-boot-maven-plugin` which does not recognize it anymore, when one liner. I will define explicitly the main class in its configuration section, it works like that

TheBestPessimist replied:
I think that's a good practice anyway (defining the main manually), as what should happen if there are multiple `main` functions? In a project there can be multiple `main`s for any number of reasons

enescualexandru replied:
Totally agree

TheBestPessimist commented on src/test/resources/application.properties:
```diff
@@ -0,0 +1,13 @@
+spring.datasource.username=sa
+spring.datasource.password=
+spring.datasource.driverClassName=org.h2.Driver
+spring.jpa.database-platform=org.hibernate.dialect.H2Dialect
```
i think the dialect should be postgres, since that's what we use in main src

enescualexandru replied:
Nope, I am using in-memory h2 for testing

TheBestPessimist commented on src/test/kotlin/com/shortener/ShortenerApplicationTests.kt:
```diff
@@ -0,0 +1,12 @@
+import org.junit.jupiter.api.Test
+import org.springframework.boot.test.context.SpringBootTest
+
+@SpringBootTest
```
Isn't this test useless as `UrlServiceTest` is also a `@SpringBootTest` ?

enescualexandru replied:
Is completely useless, I forgot about it. Will be deleted

TheBestPessimist commented on src/main/kotlin/com/shortener/dto/UrlShortenRequest.kt:
```diff
@@ -0,0 +1,3 @@
+package com.shortener.dto
+
+class UrlShortenRequest(var longUrl: String = "")
```
should be a data class; can `longUrl` be `val`?

enescualexandru replied:
Corrected

TheBestPessimist replied:
Not fully: the class should also be a `data class`

enescualexandru replied:
Corrected, again

TheBestPessimist commented on src/main/kotlin/com/shortener/dto/UrlShortenResponse.kt:
```diff
@@ -0,0 +1,3 @@
+package com.shortener.dto
+
+class UrlShortenResponse(var shortenedUrl: String)
```
should be a data class; can `shortenedUrl` be `val`?

TheBestPessimist replied:
Same as above: the class should be a `data class`

enescualexandru replied:
It is now

TheBestPessimist commented on src/main/kotlin/com/shortener/service/UrlServiceImpl.kt:
```diff
@@ -0,0 +1,101 @@
+        return urlEntry.getExpiresAt()!!.isAfter(LocalDateTime.now())
+    }
+
+    private fun addDefaultSchemaIfMissing(requestUrl: String): String {
```
i disagree with this implementation.

Basically, if i enter the URI `mailto:a@b.c`, the shortner will break my email and instead return `http://mailto:a@b.c`

I will have to re-read the readme to see if that is in spec.

enescualexandru replied:
either `mailto:a@b.c` or `http://mailto:a@b.c` are not valid URL's. The application is intended to shorten only URL's, not the broader URI's

enescualexandru replied:
indeed, not clearly specified in the docs

TheBestPessimist commented on src/main/kotlin/com/shortener/service/UrlServiceImpl.kt:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    private fun isValidUrl(requestUrl: String): Boolean {
+        return UrlValidator(ALLOW_LOCAL_URLS).isValid(requestUrl)
```
perf: Maybe `UrlValidator(ALLOW_LOCAL_URLS)` can be extracted to a private property

TheBestPessimist commented on src/main/kotlin/com/shortener/service/UrlServiceImpl.kt:
```diff
@@ -0,0 +1,101 @@
+            throw InvalidInputUrl("The URL you provided is invalid")
+        }
+
+        val urlEntry = UrlEntry()
```
the kotlin way of doing this initialization is
```kotlin
val urlEntry = UrlEntry().apply {
	longUrl = fixedRequestUrl
    createdAt = LocalDateTime.now()
}
```

enescualexandru replied:
I am convinced this will work if those properties were not private

TheBestPessimist commented on src/main/kotlin/com/shortener/domain/UrlEntry.kt:
```diff
@@ -0,0 +1,61 @@
+import javax.validation.constraints.Size
+
+@Entity(name = "URL_ENTRIES")
+class UrlEntry {
```
there is no equals and hashCode here. that worries me

TheBestPessimist commented on src/main/kotlin/com/shortener/service/UrlServiceImpl.kt:
```diff
@@ -0,0 +1,101 @@
+        return UrlShortenResponse(url)
+    }
+
+    override fun decodeUrl(encodedSequence: String): String {
```
This thing looks bad performance wise imo: 

If i were to read the same short url (into a long url) `1 000 000 000 000` times, we would have to call the decode function every time.

Why not persist the shortened URL with every entry as well? 

Right now we hit the database and also hit the CPU.

Secondary to this: what happens when we change the url encoder? My gut feeling tells me we will not be able to find the long url by the short one, as the decoder will not return the correct data. This is very wrong ❗

georgemarin replied:
I dont really care about performance or if the application is correct or not, but something cool that you might want to look into is Hazelcast to cache and avoid rendundant database calls.

P.S: Hazelcast is really easy to add and you dont need a different cluster for it, the node will run in the same JVM as your application.
P.S2: if you consider making your entries immutable you dont necessarily need to use something as complex as hazelcast, you can use a simpler import com.github.benmanes.caffeine.cache.Caffeine cache

just some random thoughts on db caching that I have used in the past that you might find intresting.

enescualexandru replied:
Thanks for the caching suggestions. I am currently working on it, in different branch. Decided to use `redis`

TheBestPessimist commented on src/main/kotlin/com/shortener/service/UrlServiceImpl.kt:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    private fun addBaseUrlToEncodedSequence(sequence: String): String {
+        val baseUrl = ServletUriComponentsBuilder.fromCurrentContextPath().build().toUriString()
```
`baseUrl` can be class property

enescualexandru replied:
Done

TheBestPessimist commented on src/main/kotlin/com/shortener/domain/UrlEntryRepository.kt:
```diff
@@ -0,0 +1,9 @@
+
+@Repository
+interface UrlEntryRepository : JpaRepository<UrlEntry, Long> {
+    fun findAllByOrderByIdDesc(): List<UrlEntry>
```
Name of `findAllByOrderBy`: what means `findAllBy`? By what?

Is that a typo, because i don't understand what the function is supposed to do.

enescualexandru replied:
Is not a typo. This is how the jpa is deriving the query from the method name. By not specifying the attribute(`findAllBy[attribute]OrderByIdDesc`, I am expecting no `where` clause to be added, therefore, all entries will be fetched, reversed ordered by id. Note: `findAllOrderByIdDesc` does not work, already tried. This method will be certainly refactored, once we'll have users and gain the ability to target the entries

georgemarin replied:
You can jus use a Sort parameter instead of hard coding the value in this function

enescualexandru replied:
And why should I then provide this Sort parameter whenever I am calling the function, when I don't need it ?

TheBestPessimist commented on src/main/kotlin/com/shortener/exception/InvalidInputUrl.kt:
```diff
@@ -0,0 +1,7 @@
+package com.shortener.exception
+
+class InvalidInputUrl : RuntimeException {
```
Note: in kotlin everything is a RuntimeException. In other words, there are no "checked exceptions". 

So maybe it makes sense to extend `Exception` instead of `RuntimeException`

enescualexandru replied:
Done, `Exception` it is

TheBestPessimist commented on src/test/kotlin/com/shortener/service/UrlServiceTest.kt:
```diff
@@ -0,0 +1,175 @@
+    fun getAllShortenedUrlsTest() {
+        val entry1 = createUrlEntry(goodUrl, false)
+        val entry2 = createUrlEntry(goodUrlNoSchema, false)
+        urlEntryRepository.saveAll(arrayListOf(entry1, entry2))
```
note: there is also the function `listOf` instead of `arrayListOf`.

enescualexandru replied:
Replaced

TheBestPessimist commented on src/main/kotlin/com/shortener/controller/UrlController.kt:
```diff
@@ -0,0 +1,49 @@
+class UrlController(private val urlService: UrlService) {
+
+    @GetMapping
+    fun getEncodeUrlPage(model: Model): String {
```
I will point out that i have no idea what `Model` is, so i cannot verify this

georgemarin commented on src/main/kotlin/com/shortener/controller/UrlController.kt:
```diff
@@ -0,0 +1,49 @@
+import java.net.URI
+
+
+@Controller
```
does this work without @ResponseBody? I think you want @RestController here.

P.S: Although you have a single controller I would add a mapping

enescualexandru replied:
This is not a rest api

georgemarin commented on src/main/kotlin/com/shortener/controller/UrlController.kt:
```diff
@@ -0,0 +1,49 @@
+        //TODO: remove it (displays the shortened url history)
+        model["allEncodedUrls"] = urlService.getAllShortenedUrls()
+
+        return "encode-url"
```
Magic string. perhaps use a const instead?. P.S: applicable to all strings

enescualexandru replied:
Extracted to a constant

georgemarin commented on src/main/kotlin/com/shortener/domain/UrlEntry.kt:
```diff
@@ -0,0 +1,61 @@
+    @Column(name = "created_at")
+    @DateTimeFormat(iso = DateTimeFormat.ISO.DATE_TIME)
+    @NotNull
+    private var createdAt: LocalDateTime? = null
```
making all of them null by default is really a bad idea imho. (you even use the notNull annotation). Why not use a proper value? 

P.S: look into the billion dollar mistake, the way you use it is for simplicity, and not because null is an actuall accepted value.
https://en.wikipedia.org/wiki/Tony_Hoare

enescualexandru replied:
ok, and what should be the proper value for id, for example ?

georgemarin commented on src/main/kotlin/com/shortener/domain/UrlEntry.kt:
```diff
@@ -0,0 +1,61 @@
+    @NotNull
+    private var expiresAt: LocalDateTime? = null
+
+    fun getId(): Long? {
```
What's the point of private properties and adding getters and setters? This is just an anti-pattern that was common in java (and unfortunately I don't think you could have avoided it when using frameworks). I think JPA with kotlin works fine if you remove the private access modifier as well as getters and setters.

georgemarin replied:
P.S: for get you could have just inlined the functions.  fun getLongUrl() =  longUrl

enescualexandru replied:
I'm into Kotlin for a few days only, still learning a lot of things. I don't know yet what is good or bad, or customary, or antipatterns or whatever, in this language. Asking "why's" and "what's" on Kotlin specific stuff will not get you answers from me. 

Suggestions are always welcomed.

georgemarin commented on src/main/kotlin/com/shortener/exception/ExceptionHandler.kt:
```diff
@@ -0,0 +1,23 @@
+class ExceptionHandler : ResponseEntityExceptionHandler() {
+
+    @ExceptionHandler(Exception::class, InvalidInputUrl::class)
+    fun handleError(req: HttpServletRequest, ex: Exception): ModelAndView? {
```
when is this function returning null so that the ModelAndView is nullable?

enescualexandru replied:
Well, never. Didn't know what I was doing

georgemarin commented on src/main/kotlin/com/shortener/service/UrlServiceImpl.kt:
```diff
@@ -0,0 +1,101 @@
+        return allShortenedUrls
+    }
+
+    private fun getNoUserUrlExpiresDays(): Long {
```
inline

enescualexandru replied:
Inline it is

georgemarin commented on src/test/kotlin/com/shortener/service/UrlServiceTest.kt:
```diff
@@ -0,0 +1,175 @@
+@ExtendWith(MockitoExtension::class)
+@SpringBootTest
+@Transactional
+class UrlServiceTest(
```
https://mockk.io/ I like this for kotlin more than mockito

enescualexandru replied:
Will use it next time, as for now I've excluded mockito from the tests, don't need it anymore

georgemarin commented on src/test/kotlin/com/shortener/service/UrlServiceTest.kt:
```diff
@@ -0,0 +1,175 @@
+     * Verifies that a shortened url which has expired cannot be converted, and exception is thrown
+     */
+    @Test
+    fun decodeUrlExpiredUrlTest() {
```
![image](https://user-images.githubusercontent.com/18738218/131987535-b828df69-42df-45df-8d46-24fdcf97ff60.png) and you can delete the comments

enescualexandru replied:
Done

enescualexandru committed:
Addressed code review comments

TheBestPessimist committed:
Replace Optional with nullable type

TheBestPessimist committed:
Style

TheBestPessimist committed:
Use non-null as much as possible in entities

enescualexandru committed:
Made the main function as one liner

enescualexandru committed:
Ensured the service method 'shortenUrl' is transactional; corrected the import in the tests

enescualexandru committed:
Added explicit column length, where required

enescualexandru committed:
Improvements

- amended the readme file, with some clarifications
- corrected the constructor for hashids, enabling a minimum hash length of 4
- changed the docker file, added eclipse temurin as the base image for the app, adoptopenjdk being as of now, deprecated

enescualexandru committed:
Changed the dto classes to data classes
